# BASIC NEURAL NETWORK WITH TENSORFLOW

## IMPORT LIBRARY

In [1]:
import tensorflow as tf # library dalam python untuk deep learning
import numpy as np # library untuk operasi perhitungan numerik
from tensorflow import keras # keras adalah API untuk tensorflow yang mempermudah dalam membuat model deep learning

2023-11-09 12:08:35.924511: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-11-09 12:08:36.510708: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-11-09 12:08:36.516472: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-11-09 12:08:43.103938: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


## CREATE DATASET

In [2]:
xs = np.array([-1.0,  0.0, 1.0, 2.0, 3.0, 4.0], dtype=float) # Fitur 
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float) # Label

# Secara logika nilai ini bisa didapatkan dengan rumus y = 2x - 1

## DEFINE A MODEL

In [3]:
# Membangun sebuah model sederhana

model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

# Model diatas memiliki 1 layer dengan 1 neuron dan 1 input shape



## COMPILE A MODEL

In [5]:
# Kompilasi Model dengan fungsi loss MSE dan optimizer sgd

model.compile(
    loss = 'mse', # Menggunakan MSE dikarenakan ini adalah masalah regresi (bukan klasifikasi)
    optimizer = 'sgd' # Menggunakan SGD (Stochastic Gradient Descent) untuk mencari nilai minimum dari fungsi loss 
)

## TRAIN A MODEL

In [8]:
# Melatih model
model.fit(xs, ys, epochs=100) # Melatih model dengan 100 iterasi (epochs) 

Epoch 1/100
1/1 [==============================] - 0s 12ms/step - loss: 0.6750
Epoch 2/100
1/1 [==============================] - 0s 9ms/step - loss: 0.6443
Epoch 3/100
1/1 [==============================] - 0s 10ms/step - loss: 0.6179
Epoch 4/100
1/1 [==============================] - 0s 10ms/step - loss: 0.5947
Epoch 5/100
1/1 [==============================] - 0s 10ms/step - loss: 0.5743
Epoch 6/100
1/1 [==============================] - 0s 10ms/step - loss: 0.5561
Epoch 7/100
1/1 [==============================] - 0s 10ms/step - loss: 0.5396
Epoch 8/100
1/1 [==============================] - 0s 12ms/step - loss: 0.5245
Epoch 9/100
1/1 [==============================] - 0s 10ms/step - loss: 0.5106
Epoch 10/100
1/1 [==============================] - 0s 14ms/step - loss: 0.4977
Epoch 11/100
1/1 [==============================] - 0s 11ms/step - loss: 0.4855
Epoch 12/100
1/1 [==============================] - 0s 10ms/step - loss: 0.4740
Epoch 13/100
1/1 [==============================] 

## INFERENCE

In [9]:
print(model.predict([10.0])) # Memprediksi nilai label dari fitur 10.0

1/1 [==============================] - 0s 84ms/step
[[18.199356]]


## KESIMPULAN

<p> Pada saat pengujian anda pasti berpikir hasil dari nilai label pada fitur 10.0 adalah 19, namun hasilnya adalah 18.199356. Hal ini dikarenakan pada saat proses training, model machine learning tidak akan pernah mendapatkan hasil yang sama persis dengan nilai label yang sebenarnya. Perlu diingat bahwa hasil yang didapati itu bersifat probabilitas, bukan nilai pasti.
</p>